### PR TITLE
Change shifts to take u32 as their second argument

### PR DESCRIPTION
**Pull request by Schaeff**
Todo:
- [x] More tests
- [x] Documentation

### CODE REVIEW

**Review by dark64 — DISMISSED** [review]

File: zokrates_core/src/static_analysis/uint_optimizer.rs [hunk 1]
```diff
@@ -280,43 +280,60 @@ impl<'ast, T: Field> Folder<'ast, T> for UintOptimizer<'ast, T> {
                     .with_max(range_max)
             }
             LeftShift(box e, box by) => {
-                // reduce the two terms
+                // reduce the left term
```
Comment (Schaeff) [inline zokrates_core/src/static_analysis/uint_optimizer.rs]:
```suggestion
                // reduce both terms
```

File: zokrates_core/src/static_analysis/uint_optimizer.rs [hunk 2]
```diff
@@ -280,43 +280,60 @@ impl<'ast, T: Field> Folder<'ast, T> for UintOptimizer<'ast, T> {
...
+                let e_max: usize = e
+                    .metadata
+                    .clone()
+                    .unwrap()
+                    .max
+                    .to_dec_string()
+                    .parse()
+                    .unwrap();
+
+                println!("e_max: {:x?}, by_u: {}", e_max, by_u);
```
Comment (Schaeff) [inline zokrates_core/src/static_analysis/uint_optimizer.rs]:
```suggestion
```

File: zokrates_core/src/static_analysis/uint_optimizer.rs [hunk 3]
```diff
@@ -280,43 +280,60 @@ impl<'ast, T: Field> Folder<'ast, T> for UintOptimizer<'ast, T> {
...
+
+                let max = (e_max & (2_usize.pow(range as u32) - 1)) >> by_u;
+
+                println!("MAX {:x?}", max);
```
Comment (Schaeff) [inline zokrates_core/src/static_analysis/uint_optimizer.rs]:
```suggestion
```

**Review by Schaeff — COMMENTED** [review]
Noticed some issues

**Comment by Schaeff** [comment]
@dark64 can you have another look as I found some issues and changed some things?

**Comment by dark64** [comment]
> @dark64 can you have another look as I found some issues and changed some things?

I see, looks good to me!

**Review by dark64 — APPROVED** [review]

**Pull request merged**